util/src/main/java/io/kubernetes/client/informer/cache/DeltaFIFO.java

```diff
@@ -0,0 +1,505 @@
+        this.queueActionLocked(Deleted, new DeletedFinalStateUnknown<>(knownKey, deletedObj));
+      }
+
+      if (!this.populated) {
+        this.populated = true;
+        this.initialPopulationCount = list.size() + queueDeletion;
+      }
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  /** Resync the delta FIFO. */
+  @Override
+  public void resync() {
+    lock.writeLock().lock();
+    try {
+      if (this.knownObjects == null) {
+        return;
+      }
+
+      List<String> keys = this.knownObjects.listKeys();
+      for (String key : keys) {
+        syncKeyLocked(key);
+      }
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * List keys list.
+   *
+   * @return the list
+   */
+  @Override
+  public List<String> listKeys() {
+    lock.readLock().lock();
+    try {
+      List<String> keyList = new ArrayList<>(items.size());
+      for (Map.Entry<String, Deltas<ApiType>> entry : items.entrySet()) {
+        keyList.add(entry.getKey());
+      }
+      return keyList;
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Get object.
+   *
+   * @param obj the obj
+   * @return the object
+   */
+  @Override
+  public Object get(Object obj) {
+    String key = this.keyOf(obj);
+    return this.getByKey(key);
+  }
+
+  /**
+   * Gets get by key.
+   *
+   * @param key the key
+   * @return the get by key
+   */
+  @Override
+  public Deltas<ApiType> getByKey(String key) {
+    lock.readLock().lock();
+    try {
+      Deltas deltas = this.items.get(key);
+      if (deltas != null) {
+        // TODO: make a deep copy
+        ArrayList<Deltas.Delta> copiedDeltas = new ArrayList<>(deltas.getDeltas());
+        return new Deltas(copiedDeltas);
+      }
+    } finally {
+      lock.readLock().unlock();
+    }
+    return null;
+  }
+
+  /**
+   * List list.
+   *
+   * @return the list
+   */
+  @Override
+  public List<Object> list() {
+    lock.readLock().lock();
+    List<Object> objects = new ArrayList<>();
+    try {
+      // TODO: make a generic deep copy utility
+      for (Map.Entry<String, Deltas<ApiType>> entry : items.entrySet()) {
+        ArrayList<Deltas.Delta<ApiType>> copiedDeltas =
+            new ArrayList<>(entry.getValue().getDeltas());
+        Deltas<ApiType> newDeltas = new Deltas(copiedDeltas);
+        objects.add(newDeltas);
+      }
+    } finally {
+      lock.readLock().unlock();
+    }
+    return objects;
+  }
+
+  /**
+   * Pop deltas.
+   *
+   * @param func the func
+   * @return the deltas
+   * @throws Exception the exception
+   */
+  @Override
+  public Deltas<Object> pop(PopProcessFunc<Deltas<Object>> func) throws InterruptedException {
+    lock.writeLock().lock();
+    try {
+      while (true) {
+        while (queue.isEmpty()) {
+          notEmpty.await();
+        }
+        // there should have data now
+        String id = this.queue.removeFirst();
+        boolean exist = this.items.containsKey(id);
```

Comment:
Move this down to the conditional and inline.

Reply:
check 👍🏻